api/src/models/user/user.methods.ts

```diff
@@ -1 +1,89 @@
-// Placeholder for future methods
+import { IUser, IUserDocument,
+  IUserModel, IUserRegistrationDetails } from "./user.types";
+import bcrypt from "bcryptjs";
+import { decrypt } from "../../utils/crypto";
+/**
+ * Find user by googleId, if not found, create user, populating with google
+ * profile data
+ * @param this *
+ * @param data User data
+ */
+export async function findOneOrCreateByGoogleId(this: IUserModel, data: IUser): Promise<IUserDocument> {
+  const documents = await this.find({ "auth.googleId": data.auth.googleId });
+  if (documents && documents.length > 0) {
+    return documents[0];
+  }
+  // If not found, create
+  try {
+    const newUserEntry = await this.create(data);
+    return newUserEntry;
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+/**
+ * Finds user by google id
+ * @param this
+ * @param id google id
+ */
+export async function findByGoogleId(this: IUserModel, id: string): Promise<IUserDocument> {
+  try {
+    return await this.findOne({ "auth.googleId": id});
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+/**
+ *  Registers user. Checks to ensure e-mail address is unique
+ */
+export async function registerUser(this: IUserModel, details: IUserRegistrationDetails) {
+    const users = await this.findByEncryptedEmail(details.encryptedEmail); // This won't work
+    if (users && users.length > 0) {
+      throw new Error(`User with email ${decrypt(details.encryptedEmail)} already exists`);
+    } else {
+      // Create
+      const hashedPassword = await bcrypt.hash( details.plainTextPassword, 10);
+      const newUser = await this.create({
+        firstName: details.firstName,
+        lastName: details.lastName,
+        auth: {
+          email: details.encryptedEmail,
+          password: hashedPassword
+        },
+        avatar: [{ url: "defaultAvatar"}],
+        connections: {},
+        connectionOf: {},
+        threads: {
+          started: {},
+          commented: {},
+          liked: {},
+          shared: {}
+        }
+      });
+      return newUser;
+    }
+}
+
+export async function findByEncryptedEmail (this: IUserModel, encryptedEmail: string) {
+  const decryptedEmail = decrypt(encryptedEmail);
+  const allRecords = await this.find();
+  return allRecords.filter((records) => {
+    return decrypt(records.auth.email) === decryptedEmail;
+  });
```

Comment:
This  and the `findOneByEncryptedEmail ` function both seem hacky, and there could be a better way to do this. 
Because we're encrypting e-mails, there's no built-in mongo way to search a record by e-mail since the e-mail is stored as random encrypted characters.

My solution was to get all the users in the db, which converts them into a JS array, where I can then perform array filter functions to decrypt and find the correct e-mail to match.

Reply:
With the same passphrase are encryptions unique? could we encrypt the query email and search for that?

Reply:
Believe it or not, I tested this very question; encryptions are unique. The same string value encrypted with the AES function produces a different cipher. 

Doing some research, it looks like this by design (and good security practice).

Reply:
I guess that makes sense. This will be fine for now, we can come back to this later and decide if we need a change or not, I guess the other way could be to also include a hash for searching, though I can imagine a hash and encryption side by side is going to make the encryption much less secure. Nice work on this BTW.